LLVM port from release/13.x to release_13x

abrahamtovarmob opened the pull request:
Port from LLVM release/13.x. Commits from release_12x (specially CI changes) added on top of the commits. Final commit is mainly changes that required manual merge in order to build the project.
Build procedure is same as release_12x. Test for flang run successfully but some test need to be modified, flang PR also open for review

schweitzpgi committed:
Add DebugInfo tests

Cherry-picked cabccf8f2d00daf1d740b18a1c80e9b5f4593f60.

bryanpkc committed:
Port LLVM changes from release_11x

This is a combination of cherry-picking the following patches:

Commit 82b38d24a518:

[ClassicFlang] Port release_90 changes from flang-compiler/llvm

Cherry-picked commit 2085211cfcca70411dc63f0d08763facc8a02090 by Eric Schweitz,
resolved merge conflicts, fixed build failures (e.g. adapted CGDebugInfo.cpp to
the new API), and fixed the DIGlobalVariable unit tests, which have been broken
since commit edfad65eebdf045b050f37380b6b61d673513982.

Commit 885dd87e5fdc:

[DebugInfo] Removal of DIFortranArrayType and DIFortranSubrange

These extensions are no more required after merge of below PR.

Commit 5c9b2e0867d5:

Modification to incorporate comments from @bryanpkc.

Commit 822de2c2a9cc:

[AsmPrinter] Fix redundant names and types

A bug was introduced in 82b38d24a51871a210a184dd558e9b955a421e8e while
cherry-picking a DIGlobalVariable-related patch.

Port Classic Flang to LLVM 13

File changes to TargetLibraryInfo and DebugLocEntry to adapt to the code from
release/13.x and make it work. Comment out the changes due a segmentation
fault, code need to be reviewed properly once all commits are in

Port driver changes from release/11x

Cherry-picked c51f89679135f84675f492d560ec5535c2000cfe by Varun Jayathirtha
from release_90, and resolved merge conflicts.

To avoid conflicts with the new Flang, lib/Driver/ToolChains/Flang.{cpp,h}
have been renamed to ClassicFlang.{cpp,h}, and the ENABLE_CLASSIC_FLANG macro
is introduced to select which incarnation of Flang to build. The macro is set
by running CMake with -DLLVM_ENABLE_CLASSIC_FLANG.

After merge with LLVM 11: Move flang options to the end of the definitions list.

Port Classic Flang to LLVM 13

File changes to port from the release_12x to the release_13x.

bryanpkc committed:
Fix use of classic Flang as preprocessor

Port Classic Flang to LLVM 13

LLVM por from release_12x to release_13x, changes done in order to
make project able to be build.

bryanpkc committed:
Fix test failures when in classic Flang mode

Add a new lit feature tag "classic_flang" to select which tests can or cannot be
run when the driver is built for classic Flang. Handle LLVM_ENABLE_CLASSIC_FLANG
in llvm/cmake/modules/HandleLLVMOptions.cmake instead of clang/CMakeLists.txt so
that macro works in both clang and llvm.

bryanpkc committed:
Merge FLANG_LLVM_EXTENSIONS macro into ENABLE_CLASSIC_FLANG

alokkrsharma committed:
Support for DWARF 4/5 and fix of issues related to -gdwarf-X options (#92)

* Support for -gdwarf-5 option in Flang driver.

Summary:
  FLANG driver doesnt pass -gdwarf-4/5 to flang1 in form of xbits,
while it passes for -gdwarf-2/3
   -gdwarf-2 => -x 120 0x200
   -gdwarf-3 => -x 120 0x4000

Due to this -gdwarf-5 is never honored and default option -gdwarf-4
is taken.
    # flang -gdwarf-5 test.f90
    # llvm-dwarfdump a.out | grep version
    0x00000000: Compile Unit: length = 0x0000008e version = 0x0004

  Now 0x1000000/0x2000000 will be passed for -gdwarf-4/5
   -gdwarf-4 => -x 120 0x1000000
   -gdwarf-5 => -x 120 0x2000000

Testing:
  - GNU gdb fortran testsuite
  - check-llvm
  - check-debuginfo

* Flang doenst choose correct dwarf version when multiple -g/-gdwarfN mentioned

 Summary:
When multiple -g/-gdwarfN options are passed together at compile time,
flang chooses the least one. Clang/gfortran etc choose the last one.

-gdwarf-5 -gdwarf-3 => flang chooses 5 while clang/gfortran choose 3
-gdwarf-5 -g => flang choses the default while clang/gfortran choose 5

  Testing:
- check-llvm
- check-debuginfo

* Default dwarf version should be 4 for -g with flang

Currently flang dumps dwarf version 2 for -g and 4 for absence of -g

-------------------------
$ flang my.f90
$ llvm-dwarfdump a.out | grep version
0x00000000: Compile Unit: length = 0x0000003d version = 0x0004 abbr_offset = 0x0000 addr_size = 0x08 (next unit at 0x00000041)

$ flang -g my.f90
$ llvm-dwarfdump a.out | grep version
0x00000000: Compile Unit: length = 0x00000047 version = 0x0002 abbr_offset = 0x0000 addr_size = 0x08 (next unit at 0x0000004b)
-------------------------

It should be 4 for -g as it is the case with clang.

bryanpkc committed:
Reduce downstream delta

This commit is motivated by reducing the merge burden by shrinking the
diff between llvm upstream and classic-flang-llvm-project.

Outside of Flang, Fortran code is fed through the Compile phase, and
the appropriate tooling is picked up through ToolChain::SelectTool.

Classic Flang introduced a FortranFrontend, but these days this seems
unnecessary. Fortran can go through the same machinery as everything else.

* Use the Preprocess phase to preprocess Fortran code. This phase is
  always combined with the Compile phase.

* Use the Compile phase to lower Fortran code to LLVM IR, and use the
  Backend phase to compile and optimize the IR. These phases are never
  combined.

* Remove FortranFrontendJobClass.

* Remove FortranFrontend tool (instead it's just the Flang tool, which
  in Classic Flang mode is Classic Flang).

* Update tests which inspect the output of the Classic Flang tooling,
  and ensures that the driver does the right thing for various types of
  inputs.

Based on a patch from Peter Waller <peter.waller@arm.com>.

kiranchandramohan committed:
Removing a few CI pipelines

Github Actions added to pre-compile artifacts for flang

Github Actions to build release_11x and release_12x

release_100 is still built as it was.
Artifacts names include the branch name.

Enable ccache for GitHub Actions

Allow full path to flang1 and flang2 in classic flang tests.

Remove unnecessary caret from path in classic-flang test

Remove the builds for gcc-9 and llvm-9

Build and test Flang project on PRs

Prepare llvm and clang for ARM64 build of Flang

Run Dockerfile script on Github's server to prepare a docker image with llvm and clang built and installed for Flang to use when testing code.

bryanpkc committed:
Fix -fveclib=PGMATH

The use of #ifdef in include/clang/Driver/Options.td was incorrect and
unsupported. As a result -fveclib=PGMATH was silently ignored, and
in LLVM 12, it causes the invocation to fail. This patch unguards the
option so that it is parsed correctly, but lets the FLANG_LLVM_EXTENSIONS
macro continue to toggle the feature.

SouraVX committed:
[DebugInfo] Support of `-gpubnames` in Driver

Removing CI test-base for release_100 (release_10x)

SouraVX committed:
[ClassicFlang][Driver] Correct the LLVM version passed by the Driver

Allow pre-compile jobs to be trigerred manually.

Add build scripts and make CI use them.

isuruf committed:
support macOS

pawosm-arm commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -1932,22 +1932,6 @@ void DwarfDebug::collectEntityInfo(DwarfCompileUnit &TheCU,
     // Finalize the entry by lowering it into a DWARF bytestream.
     for (auto &Entry : Entries)
       Entry.finalize(*Asm, List, BT, TheCU);
-    List.finalize();
```
Can you give some rationale behind this removal?

abrahamtovarmob replied:
This change comes from the original code https://github.com/llvm/llvm-project/blob/release/13.x/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp#:~:text=for%20(auto%20%26Entry,List%2C%20BT%2C%20TheCU)%3B
I pushed as-it-is

abrahamtovarmob replied:
Paul commented offline that this change comes from release_12x, I see that the issue was that the base code for release/13.x is different and when I applied the commit for this piece of code the build was breaking, hence I remove the code. What I will do now is to pick up the commit again and do the required changes for release_13x be compatible with it. I will commit-push again once is ready.

pawosm-arm commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -2693,33 +2677,6 @@ void DebugLocEntry::finalize(const AsmPrinter &AP,
     List.setTagOffset(*DwarfExpr.TagOffset);
 }
 
-inline static DbgValueLoc mkDbgValueLoc(const DIExpression *expr,
```
Again, can you write some words why this had to go?

abrahamtovarmob replied:
This change comes from the original code https://github.com/llvm/llvm-project/blob/release/13.x/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp#:~:text=List.setTagOffset(*DwarfExpr.TagOffset)%3B

abrahamtovarmob replied:
Paul commented offline that this change comes from release_12x, I see that the issue was that the base code for release/13.x is different and when I applied the commit for this piece of code the build was breaking, hence I remove the code. What I will do now is to pick up the commit again and do the required changes for release_13x be compatible with it. I will commit-push again once is ready.

abrahamtovarmob commented:
I will temporary close this PR because I'm committing changes to my local branch and I don't want them to reflect here until it's ready. I will reopen until all test are passing

abrahamtovarmob committed:
[ClassicFlang][Driver] Correct the LLVM version passed by the Driver

Change the diver version to 130 since is the correct version for
release_13x driver

LLVM port for release_13x

Change to Options.td in order to add the correct invocation for
ffixed_line_length_VALUE.

pawosm-arm commented:
@abrahamtovarmob how much 'ready' this PR is? Seems it is the pre-requisite for everything else in the LLVM13-merge epic.

kiranchandramohan commented:
@abrahamtovarmob Can you cherry-pick https://github.com/flang-compiler/classic-flang-llvm-project/pull/64 also?

bryanpkc commented:
We are doing some internal tests of this branch. We will post our feedback when the tests are done.

abrahamtovarmob commented:
> @abrahamtovarmob Can you cherry-pick #64 also?

Done

RichBarton-Arm commented:
@kiranchandramohan , @bryanpkc  - are you happy to approve this one? If it looks ok could we merge it then add additional patches on top as separate PRs if needed?

RichBarton-Arm reviewed (approved):
Happy to approve this from the Arm side.

bryanpkc commented on clang/lib/CodeGen/BackendUtil.cpp:
```diff
@@ -415,6 +415,11 @@ static TargetLibraryInfoImpl *createTLII(llvm::Triple &TargetTriple,
     TLII->addVectorizableFunctionsFromVecLib(
         TargetLibraryInfoImpl::DarwinLibSystemM);
     break;
+#ifdef FLANG_LLVM_EXTENSIONS
```
We have replaced `FLANG_LLVM_EXTENSIONS` with `ENABLE_CLASSIC_FLANG` in 56f27ac85775d523757a6826ceb98078972256e7. You probably shouldn't undo that change.

abrahamtovarmob replied:
Somehow I left out the commit, thanks for mention it

bryanpkc commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -4810,7 +4810,68 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
     }
   }
 
+#ifdef FLANG_LLVM_EXTENSIONS
```
We have replaced FLANG_LLVM_EXTENSIONS with ENABLE_CLASSIC_FLANG in 56f27ac. You probably shouldn't undo that change.

abrahamtovarmob replied:
Somehow I left out the commit, thanks for mention it

bryanpkc replied:
Fixed.

bryanpkc commented on clang/include/clang/Driver/Types.def:
```diff
@@ -72,6 +72,12 @@ TYPE("c++-module-cpp-output",    PP_CXXModule, INVALID,         "iim",    phases
+TYPE("f95-cpp-input",            F_FreeForm,   PP_F_FreeForm,   "F95",    phases::Preprocess, phases::Compile, phases::Backend, phases::Assemble, phases::Link)
+#endif
 TYPE("f95",                      PP_Fortran,   INVALID,         nullptr,  phases::Compile, phases::Backend, phases::Assemble, phases::Link)
 TYPE("f95-cpp-input",            Fortran,      PP_Fortran,      nullptr,  phases::Preprocess, phases::Compile, phases::Backend, phases::Assemble, phases::Link)
```
Lines 81 and 82 used to be disabled if `ENABLE_CLASSIC_FLANG` was set. Now these two lines are processed along with lines 76 though 79. Wouldn't they cause a conflict? A similar problem exists in `clang/lib/Driver/Types.cpp`.

bryanpkc commented on clang/lib/Driver/Types.cpp:
```diff
@@ -235,6 +264,10 @@ types::ID types::lookupTypeForExtension(llvm::StringRef Ext) {
+           .Case("f", TY_PP_F_FixedForm)
+#endif
            .Case("F", TY_Fortran)
            .Case("f", TY_PP_Fortran)
```
These two cases will never match if `ENABLE_CLASSIC_FLANG` is set, right?

bryanpkc commented on clang/lib/Driver/Types.cpp:
```diff
@@ -270,6 +303,20 @@ types::ID types::lookupTypeForExtension(llvm::StringRef Ext) {
+#endif
            .Case("F90", TY_Fortran)
            .Case("f90", TY_PP_Fortran)
            .Case("F95", TY_Fortran)
```
These cases will never match if `ENABLE_CLASSIC_FLANG` is set, right?

abrahamtovarmob replied:
I added the #else exclusion, now these "Case" with TY_Fortran and TY_PP_Fortran wont be match

bryanpkc replied:
OK thanks.

bryanpkc commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -1223,11 +1223,13 @@ void DwarfDebug::beginModule(Module *M) {
     }
 
     DenseSet<DIGlobalVariable *> Processed;
+//    CU.setGlobalVarMap(&GVMap);
```
Why is this commented out, along with other large sections of code below?

bryanpkc replied:
Consider using `#if 0` instead of `//` to disable large sections of code.

abrahamtovarmob replied:
Fixed

bryanpkc commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -1853,7 +1871,10 @@ void DwarfDebug::collectEntityInfo(DwarfCompileUnit &TheCU,
 
     LexicalScope *Scope = nullptr;
-    const DILocalVariable *LocalVar = cast<DILocalVariable>(IV.first);
+	 const DILocalVariable *LocalVar = cast<DILocalVariable>(IV.first);
```
Unnecessary indentation here.

abrahamtovarmob replied:
sorry, fixed

bryanpkc commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.h:
```diff
@@ -625,6 +629,12 @@ class DwarfDebug : public DebugHandlerBase {
   void emitSectionReference(const DwarfCompileUnit &CU);
 
+  /// Populate dependent type variable map
+  void populateDependentTypeMap();
```
Didn't you comment out the definition of this function in `llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp`?

bryanpkc replied:
Fixed.

bryanpkc commented:
You seem to have deleted some test cases:

- `llvm/test/DebugInfo/Generic/fortran-subprogram-at.ll`
- `llvm/test/DebugInfo/Generic/more-subprogram-attr.ll`
- `llvm/test/DebugInfo/X86/DICommonBlock.ll`

Is that intentional?

bryanpkc commented on clang/include/clang/Driver/Options.td:
```diff
@@ -4463,49 +4477,50 @@ def Xflang : Separate<["-"], "Xflang">,
 The default is the current directory.}]>;
 
-def ffixed_form : Flag<["-"], "ffixed-form">, Group<f_Group>,
+def ffixed_form : Flag<["-"], "ffixed-form">, Group<fortran_format_Group>,
```
Question: What is the difference between `f_Group` and `fortran_format_Group` (or `gfortran_Group`)?

abrahamtovarmob replied:
f_Group is the original group name that is found in llvm/13.x. Is the Target-independent compilation options
fortran_format_Group is a group for the Fortran source format
gfortran_Group is the gfortran options that we recognize in the driver and pass along when invoking GCC to compile Fortran code.
After initial merge, flang tests fails because it cannot find the symbol, for example, ffixed_form. I don't understand the specifics of each one of it, I based on the code found on release_12x to have an idea how to use the groups names. A more complex merge could be to create a group to encapsulate all three different groups, but as I don't understand specifics of it I cannot tell if that is a good idea and ultimate solution. I vouch for the more simplest solution that was to just change the group name, as it was done between llvm release/12.x and classic release_12x

RichBarton-Arm replied:
fortran_format_Group is used to help the classic flang driver process the multiple gfortran-style and nvfort/pgfort-style options that control Fortran source format (Mfixed vs -ffixed-form, etc.) I think Abraham's merge here is correct.

In the past, upstream LLVM would have had -ffixed-form in the gfortran_Group to make clang pass it on to gfortran (or ignore it). That functionality broke upstream (I think) and anyway got replaced by new LLVM Flang's support for the options in LLVM 13. LLVM flang places -ffixed-form it in the f_Group (which I think is probably wrong, but that's an upstream bug.) The only implication of that is that the Clang driver (not Flang) will claim all unprocessed f_Group options to prevent an unused argument warning.

bryanpkc replied:
> The only implication of that is that the Clang driver (not Flang) will claim all unprocessed f_Group options to prevent an unused argument warning.

Thanks @RichBarton-Arm for this explanation. `ToolChains/ClassicFlang.cpp` processes `OPT_fortran_format_Group` but nothing in the code base uses `OPT_gfortran_Group` and `OPT_f_Group`. Could you point me to the Clang driver code that goes through options in `OPT_f_Group` in order to claim them?

RichBarton-Arm replied:
The bit of code that claims the f_Group options is in Clang's default Driver's ConstructJob function [here](https://github.com/flang-compiler/classic-flang-llvm-project/blob/f4f49d9d0d699f3ac32c1037516c9ab17551991e/clang/lib/Driver/ToolChains/Clang.cpp#L6513)). Seems a shame not to use it to e.g. organise clang's messy --help message to group similar options, but it seems this is all it's used for upstream in clang.

The gfortran support has all completely [bitrotted away](https://reviews.llvm.org/rG6a75496836ea14bcfd2f4b59d35a1cad4ac58cee) upstream in Clang. It will anyway not make much sense when LLVM Flang is  mature. We already agreed that LLVM Flang would have a gfortran-like CLI so we'll want to use these Options.td entries for LLVM Flang options eventually. We might even be able to remove the whole group now. I'll talk to Andrzej about that, but it won't have much bearing on what we're talking about here.

bryanpkc replied:
Thanks; I didn't realize that `OPT_clang_ignored_f_Group` was related to `OPT_f_Group`. I was trying to understand whether this PR needs to bother changing all the `f_Group` into `gfortran_Group` (for options used by Classic Flang), since it introduces downstream delta that may not really be necessary. If we leave them in `f_Group`, `ToolChains/ClassicFlang.cpp` should claim all of them anyway, and the fact that Clang may try to claim unused copies of such options should be irrelevant.

bryanpkc commented:
Nit: A couple of your commit messages have the wording "Port to LLVM release_12x" or "LLVM port for release_12x", which is misleading. I think it's better to say "Port Classic Flang to LLVM 13".

abrahamtovarmob commented:
> Nit: A couple of your commit messages have the wording "Port to LLVM release_12x" or "LLVM port for release_12x", which is misleading. I think it's better to say "Port Classic Flang to LLVM 13".

I had changed the name of those commits

abrahamtovarmob commented:
> You seem to have deleted some test cases:
> 
> * `llvm/test/DebugInfo/Generic/fortran-subprogram-at.ll`
> * `llvm/test/DebugInfo/Generic/more-subprogram-attr.ll`
> * `llvm/test/DebugInfo/X86/DICommonBlock.ll`
> 
> Is that intentional?

I didn't got starting commit to cherry pick, thus its omission wasn't intentional

abrahamtovarmob committed:
Port Classic Flang to LLVM 13

LLVM por from release_12x to release_13x, changes done in order to
make project able to be build.

bryanpkc commented on clang/include/clang/Basic/CodeGenOptions.h:
```diff
@@ -55,10 +55,11 @@ class CodeGenOptions : public CodeGenOptionsBase {
+    Accelerate, // Use the Accelerate framework.
+    LIBMVEC,    // GLIBC vector math library.
+    MASSV,      // IBM MASS vector library.
+    PGMATH,     // PGI math library.
```
Please keep the spacing consistent with LLVM 13.

bryanpkc commented on clang/include/clang/Frontend/Utils.h:
```diff
@@ -70,6 +73,31 @@ void InitializePreprocessor(Preprocessor &PP, const PreprocessorOptions &PPOpts,
+                                : llvm::APInt::getMaxValue(TypeWidth);
+  SmallString<4096> str;
+  MaxVal.toString(str, 10, isSigned);
+  Builder.defineMacro(MacroName, str + ValSuffix);
```
I suggest keeping this code in sync with [LLVM 13](https://github.com/llvm/llvm-project/blob/release/13.x/clang/lib/Frontend/InitPreprocessor.cpp#L171) (and current main branch), if possible:
```c++
Builder.defineMacro(MacroName, toString(MaxVal, 10, isSigned) + ValSuffix);
```

bryanpkc commented on clang/lib/Frontend/InitPreprocessor.cpp:
```diff
@@ -159,7 +159,6 @@ static void DefineFloatMacros(MacroBuilder &Builder, StringRef Prefix,
-
 /// DefineTypeSize - Emit a macro to the predefines buffer that declares a macro
 /// named MacroName with the max value for a type with width 'TypeWidth' a
 /// signedness of 'isSigned' and with a value suffix of 'ValSuffix' (e.g. LL).
```
These two `static` versions of `DefineTypeSize` should have been moved to `clang/include/clang/Frontend/Utils.h`.

bryanpkc commented on llvm/include/llvm/Analysis/TargetLibraryInfo.h:
```diff
@@ -92,6 +92,7 @@ class TargetLibraryInfoImpl {
     DarwinLibSystemM, // Use Darwin's libsystem_m.
     LIBMVEC_X86,      // GLIBC Vector Math library.
     MASSV,            // IBM MASS vector library.
+    PGMATH,     // PGI math library.
```
Please align the comments with those on surround lines.

bryanpkc commented:
Could you reorder the commits so that they are roughly in the same order as on the release_12x branch? Could you also please squash the `DwarfDebug` changes in 01f56f30430c80f04a86995ef4fad7e4a90f5570 into 885039102d331c9cd8f21c6a8ed2f914b49a785c, and the Types.def and Types.cpp changes in the same commit into 73cafd790b61ca782822be73db790a19a8e26ef4 and then into 769779faf3a1c452941d21308dc76e6467a48caa? I think that will eliminate a lot of the confusing back-and-forth changes, as well as changes that don't seem to belong together in one commit.

abrahamtovarmob commented:
> Could you reorder the commits so that they are roughly in the same order as on the release_12x branch? Could you also please squash the `DwarfDebug` changes in [01f56f3](https://github.com/flang-compiler/classic-flang-llvm-project/commit/01f56f30430c80f04a86995ef4fad7e4a90f5570) into [8850391](https://github.com/flang-compiler/classic-flang-llvm-project/commit/885039102d331c9cd8f21c6a8ed2f914b49a785c), and the Types.def and Types.cpp changes in the same commit into [73cafd7](https://github.com/flang-compiler/classic-flang-llvm-project/commit/73cafd790b61ca782822be73db790a19a8e26ef4) and then into [769779f](https://github.com/flang-compiler/classic-flang-llvm-project/commit/769779faf3a1c452941d21308dc76e6467a48caa)? I think that will eliminate a lot of the confusing back-and-forth changes, as well as changes that don't seem to belong together in one commit.

Commit order fixed and excessive back-and-forth changes squashed.

pawosm-arm commented:
Can someone having LLVM12-based (or older) classic flang check one thing for me? I've been trying to examine LLVM-IR files emitted by LLVM13-based flang (with changes from this PR applied locally). What I noticed is that no matter if -g is passed on the command line, the debug info is always present at the end of LLVM-IR file! And it happens at all of the optimization levels. Could it be that something got messed up in the frontend driver and it went unnoticed for some time?!

RichBarton-Arm commented:
@bryanpkc  - I have approved from Arm side and Shivram has approved also. If you are happy can you do the push?

pawosm-arm commented:
I guess flang part must be merged first, otherwise certain test cases will be failing.

RichBarton-Arm commented:
@pawosm-arm  - only if we enable the flang CI builds with LLVM13 at this point. We can push this then enable CI only once the Flang part has landed.

bryanpkc reviewed (approved):
I have reviewed the changes again and the updated ordering of the commits. LGTM.
